Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp

```diff
@@ -0,0 +1,419 @@
+    void visit(AST::ParameterizedType&) override;
+
+    void visit(AST::Parameter&) override;
+    // virtual void visit(StructMember&);
+    // virtual void visit(VariableQualifier&);
+
+    StringBuilder& m_stringBuilder;
+    Indentation<4> m_indent { 0 };
+};
+
+void FunctionDefinitionWriter::visit(AST::ShaderModule& shaderModule)
+{
+    AST::Visitor::visit(shaderModule);
+}
+
+void FunctionDefinitionWriter::visit(AST::FunctionDecl& functionDefinition)
+{
+    // TODO: visit return attributes
+    ASSERT(functionDefinition.maybeReturnType());
+    for (auto& attribute : functionDefinition.attributes()) {
+        checkErrorAndVisit(attribute);
+        m_stringBuilder.append(" ");
+    }
+    checkErrorAndVisit(*functionDefinition.maybeReturnType());
+    m_stringBuilder.append(" ", functionDefinition.name(), "(");
+    bool first = true;
+    for (auto& parameter : functionDefinition.parameters()) {
+        if (!first)
+            m_stringBuilder.append(", ");
+        checkErrorAndVisit(parameter);
+        first = false;
+    }
+    m_stringBuilder.append(")\n");
+    m_stringBuilder.append("{\n");
+    IndentationScope scope(m_indent);
+    checkErrorAndVisit(functionDefinition.body());
+    m_stringBuilder.append("}\n\n");
+}
+
+void FunctionDefinitionWriter::visit(AST::StructDecl& structDecl)
+{
+    // TODO: visit attributes
+    m_stringBuilder.append(m_indent, "struct ", structDecl.name(), " {\n");
+    {
+        IndentationScope scope(m_indent);
+        for (auto& member : structDecl.members()) {
+            m_stringBuilder.append(m_indent);
+            visit(member.type());
+            m_stringBuilder.append(" ", member.name());
+            for (auto &attribute : member.attributes()) {
+                m_stringBuilder.append(" ");
+                visit(attribute);
+            }
+            m_stringBuilder.append(";\n");
+        }
+    }
+    m_stringBuilder.append(m_indent, "};\n\n");
+}
+
+void FunctionDefinitionWriter::visit(AST::VariableDecl& variableDecl)
+{
+    ASSERT(variableDecl.maybeTypeDecl());
+
+    visit(*variableDecl.maybeTypeDecl());
+    m_stringBuilder.append(" ", variableDecl.name());
+    if (variableDecl.maybeInitializer()) {
+        m_stringBuilder.append(" = ");
+        visit(*variableDecl.maybeInitializer());
+    }
+}
+
+void FunctionDefinitionWriter::visit(AST::Attribute& attribute)
+{
+    AST::Visitor::visit(attribute);
+}
+
+void FunctionDefinitionWriter::visit(AST::BuiltinAttribute& builtin)
+{
+    if (builtin.name() == "vertex_index"_s) {
+        m_stringBuilder.append("[[vertex_id]]");
+        return;
+    }
+
+    if (builtin.name() == "position"_s) {
+        m_stringBuilder.append("[[position]]");
+        return;
+    }
+
+    ASSERT_NOT_REACHED();
+}
+
+void FunctionDefinitionWriter::visit(AST::StageAttribute& stage)
+{
+    switch (stage.stage()) {
+    case AST::StageAttribute::Stage::Vertex:
+        m_stringBuilder.append("vertex");
```

Comment:
The modern MSL syntax for this is `[[vertex]]`, `[[fragment]]` and `[[compute]]`.

Reply:
Which didn't actually work with something I tried.